Make log functions exportable

martinsumner opened the pull request:
To make it easier to switch to logger in kv_index_tictactree - export the log functions from leveled so that they can be reused

martinsumner committed:
Make log functions exportable

To make it easier to switch to logger in kv_index_tictactree - export the log functions from leveled so that they can be reused

ThomasArts commented on src/leveled_log.erl:
```diff
@@ -428,12 +432,15 @@ is_active_level([L|_], L, _) -> true;
+    atom(), list(), erlang:timestamp(), list(log_level()), log_base(), atom())
+        -> term().
+log_timer(LogRef, Subs, StartTime, SupportedLevels, LogBase, Tag) ->
+    {LogLevel, Log} = maps:get(LogRef, LogBase, LogBase),
```
This seems wrong. The default value for the get should either be omitted or be something computed from Supported levels or so. In any case not returning the `?LOGBASE` map and then matching that against a tuple.

martinsumner replied:
I think there should be no default (as with the log function).  If there was a default, it might have the wrong number of substitutions.  So we let it crash if there is a log to a missing log reference.

For leveled, and kv_index_tictactree they both have 100% code coverage in tests (and I maintain this as a requirement), and so there never should be a missing reference get through review.

ThomasArts commented on src/leveled_log.erl:
```diff
@@ -494,8 +501,11 @@ log_test() ->
+    ok =
+        log_timer(
+            g0001, [], os:timestamp(), [warning, error], ?LOGBASE, backend
+        ).
```
Add a test for the unknown key `wrongkey`:
`??? = log_timer(wrongkey, [], os:timestamp(), [warning, error], ?LOGBASE, backend)`
I think it should return ok and have no side effect action, similar to when the key would be present, but has a log level one is not interested in.

martinsumner replied:
See above.  I've removed the default, and so wrongkey will crash.  I've added a test to assertException to make this explicit.

ThomasArts commented on src/leveled_log.erl:
```diff
@@ -494,8 +501,11 @@ log_test() ->
+            g0001, [], os:timestamp(), [warning, error], ?LOGBASE, backend
+        ).
 
 shouldilog_test() ->
```
This tests that the function does not crash. but hardly that you get debug logs?

martinsumner replied:
I've removed the debug log test, the shouldilog_test/0 covers this functionality.

ThomasArts reviewed (commented):
One possibility to crash the log timer... if that is on purpose to find wrong log codes in testing, then fine, but could be more explicit in code

martinsumner committed:
Changes post review

Added brief description of the module to explain why the approach to logging is used.

martinsumner committed:
Merge branch 'develop-3.4' into mas-d34-openriak.i10-logger

martinsumner committed:
Result of log should be `ok`
